dynamiqs/solvers/propagator.py

```diff
@@ -28,20 +29,38 @@ def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
 
         # check that Hamiltonian is time-independent
-        if not isinstance(self.H, ConstantTimeTensor):
+        if not isinstance(self.H, (ConstantTimeTensor, PWCTimeTensor)):
             raise TypeError(
-                'Solver `Propagator` requires a time-independent Hamiltonian.'
+                f'Solver {obj_type_str(self.options.solver)} requires a'
+                ' time-independent or piecewise-constant Hamiltonian (types'
+                ' `ArrayLike`, `ConstantTimeTensor` or `PWCTimeTensor`), but the'
+                f' passed Hamiltonian has type {obj_type_str(self.H)}.'
             )
-        self.H = self.H(self.t0)
+        if isinstance(self.H, PWCTimeTensor):
+            # We build a custom times array `tstop_new` gathering `self.tstop` and
+            # `self.H.times`. This ensures that each PWC part of the Hamiltonian is
+            # taken into account during the evolution.
+
+            # keep only `H.times` values in the interval [t0, tsave[-1])
+            mask = (self.t0 <= self.H.times) & (self.H.times < self.tsave[-1])
+            extra_tstop = self.H.times[mask]
+            extra_tstop = extra_tstop.numpy(force=True)
+
+            # build new times to stop
+            self.tstop_new = np.unique(np.concatenate((self.tstop, extra_tstop)))
```

Comment:
Maybe we should call this `self.tstop` as well no? The `_new` doesn't add much to the variable name in terms of meaning IMO.

Reply:
We can't, it will override the base class `self.tstop` name that may be used internally.

Reply:
Actually I didn't notice this was only in the propagator file. I guess it's fine then. How about `self.tstop_pwc` as a name then? Or something else that carries more meaning.

Reply:
Still, I think it should be fine to override `self.tstop` no? At least if it's not, it's how it should be IMO. `self.tstop` should only carry information about when we require to stop. Here, we need to stop more often than before, but it shouldn't change the rest of the process.

Reply:
You can look at it two ways:
- either `self.tstop` is a read-only property of the base `Solver` class which denotes when `self.save(y)` should be called, and subclasses can do whatever they want on top with how they want to manage time (e.g. fixed solvers stop every dt, adaptive solvers stop often, propagator stop every pwc times, etc...)
- either `self.tstop` is a mutable property of the base class which denotes when the specific solver stops, subclasses can modify it but they need to call `self.save(y)` at the times of the original `self.tstop`

I wrote the base class with the first case in mind, I think it's much easier not to tamper with the internal of the base class (especially with the double saving `tsave`, `tmeas`, and the `tstop_counter`). I agree that it works to override `self.tstop` here, but it's really a hack IMO (for example `tstop_counter` in the base class becomes completely unrelated to `tstop`). Maybe what's confusing is that we should rename the base method `save` to `stop` to make this more explicit (for example we could stop in the base solver to log GPU usage or whatever, for now we stop to save so we called it `save`).